fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,226 @@
+            device: Device to be used for training.
+            loss_meter_type: Type of loss meter to be used.
+            metric_meter_type: Type of metric meter to be used.
+            checkpointer: Checkpointer to be used for checkpointing.
+            temperature: Temperature to be used for contrastive loss.
+            perFCL_loss_weights: Weights to be used for perFCL loss.
+            Each value associate with one of two contrastive losses in perFCL loss.
+            cos_sim_loss_weight: Weight to be used for cosine similarity loss.
+            contrastive_loss_weight: Weight to be used for contrastive loss.
+        """
+        self.perFCL_loss_weights = perFCL_loss_weights
+        self.cos_sim_loss_weight = cos_sim_loss_weight
+        self.contrastive_loss_weight = contrastive_loss_weight
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1)
+        self.ce_criterion = torch.nn.CrossEntropyLoss()
+        self.temperature = temperature
+
+        # Need to save previous local module, global module and aggregated global module at each communication round
+        # to compute contrastive loss.
+        self.old_local_module: torch.nn.Module
+        self.old_global_module: torch.nn.Module
+        self.aggregated_global_module: torch.nn.Module
+
     def get_parameter_exchanger(self, config: Config) -> ParameterExchanger:
         assert isinstance(self.model, FendaModel)
         return FixedLayerExchanger(self.model.layers_to_exchange())
+
+    def predict(self, input: torch.Tensor) -> Dict[str, torch.Tensor]:
+
+        preds = self.model(input)
+
+        if self.contrastive_loss_weight or self.perFCL_loss_weights:
+            preds["old_local_features"] = self.old_local_module.forward(input).view(len(preds["local_features"]), -1)
+            if self.perFCL_loss_weights:
+                preds["old_global_features"] = self.old_global_module.forward(input).reshape(
+                    len(preds["global_features"]), -1
+                )
+                preds["aggregated_global_features"] = self.aggregated_global_module.forward(input).reshape(
+                    len(preds["global_features"]), -1
+                )
+        return preds
+
+    def set_parameters(self, parameters: NDArrays, config: Config) -> None:
+
+        # Save the parameters of the old model
+        assert isinstance(self.model.local_module, torch.nn.Module)
+        assert isinstance(self.model.global_module, torch.nn.Module)
+        if self.contrastive_loss_weight or self.perFCL_loss_weights:
+            self.old_local_module = self.clone_model(self.model.local_module)
+            self.old_global_module = self.clone_model(self.model.global_module)
+
+        # Set the parameters of the model
+        super().set_parameters(parameters, config)
+
+        # Save the parameters of the global model
+        if self.perFCL_loss_weights:
+            assert isinstance(self.model.global_module, torch.nn.Module)
```

Comment:
I think if we do the `assert isinstance(self.model, FendaModel)`, this assert could also be removed.